File: victor/react-native/modules/elpian-widgets/android/src/main/java/expo/modules/elpianwidgets/WidgetController.kt
package expo.modules.elpianwidgets

import android.content.Context
import android.graphics.Color
import android.text.Editable
import android.text.TextWatcher
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.CompoundButton
import android.widget.EditText
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.SeekBar
import android.widget.TextView
import androidx.appcompat.widget.SwitchCompat
import androidx.core.widget.NestedScrollView
import com.google.android.flexbox.AlignItems
import com.google.android.flexbox.FlexDirection
import com.google.android.flexbox.FlexWrap
import com.google.android.flexbox.FlexboxLayout
import com.google.android.flexbox.JustifyContent
import org.json.JSONArray
import org.json.JSONObject

// Maps the VM's rn.op stream to a real android.view.View tree — no React. One
// controller per VictorSurfaceView. Flex containers use FlexboxLayout so the
// guest's flexbox styling lays out natively; the 15 widget KINDs (resolved on
// the JS side and sent as `k`) each map to a native widget. Events set native
// listeners that enqueue back to the VM via the host.
class WidgetController(
  private val context: Context,
  private val root: FlexboxLayout,
  private val host: VictorSurfaceView,
) {
  private class Entry(
    val view: View,
    val content: ViewGroup, // where children attach (== view unless a scroller)
    val kind: String,
    val props: MutableMap<String, Any?> = HashMap(),
    val events: MutableMap<String, Long> = HashMap(),
  )

  // Widget ids are 64-bit handles: the VM manager packs the owning VM id into the
  // high 32 bits (encode_handle => 2^32*vm + local), so ids routinely exceed
  // Int range. Parsing them as Int would truncate (2^32+9 -> 9) and desync the
  // native id from the JS side, so every id is a Long end to end.
  private val entries = HashMap<Long, Entry>()
  private val density = context.resources.displayMetrics.density

  private fun dp(v: Double): Int = (v * density).toInt()

  // Apply one op message (see NativeWidgetRenderer.send).
  fun apply(op: JSONObject) {
    when (op.optString("t")) {
      "create" -> create(op.getLong("id"), op.optString("cls"), op.optString("k", "view"))
      "set" -> setProp(op.getLong("id"), op.optString("k"), op.opt("v"))
      "connect" -> connect(op.getLong("id"), op.optString("e"), op.optLong("cb"))
      "disconnect" -> entries[op.getLong("id")]?.events?.remove(op.optString("e"))
      "add" -> addChild(op.getLong("p"), op.getLong("c"), op.optInt("i", -1))
      "remove" -> removeChild(op.getLong("p"), op.getLong("c"))
      "clear" -> entries[op.getLong("p")]?.content?.removeAllViews()
      "free" -> entries.remove(op.getLong("id"))
      "root" -> setRoot(op.getLong("id"))
      "toast" -> android.widget.Toast.makeText(context, op.optString("m"), android.widget.Toast.LENGTH_SHORT).show()
    }
  }

  private fun create(id: Long, cls: String, kind: String) {
    val (view, content) = build(kind)
    entries[id] = Entry(view, content, kind)
  }

  private fun build(kind: String): Pair<View, ViewGroup> {
    return when (kind) {
      "text" -> {
        val t = TextView(context)
        Pair(t, root) // text is a leaf; children (nested Text) are rare — flatten
      }
      "input" -> {
        val e = EditText(context)
        Pair(e, root)
      }
      "switch" -> Pair(SwitchCompat(context), root)
      "slider" -> Pair(SeekBar(context), root)
      "button", "victorButton" -> Pair(Button(context), root)
      "activity" -> Pair(ProgressBar(context), root)
      "image" -> Pair(ImageView(context), root)
      "scene3d" -> Pair(View(context), root) // GodotView is attached by the Godot module
      "scroll" -> {
        val sv = NestedScrollView(context)
        val inner = FlexboxLayout(context).apply { flexDirection = FlexDirection.COLUMN }
        sv.addView(inner, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        Pair(sv, inner)
      }
      else -> { // view, list, imageBackground, status, refresh
        val f = FlexboxLayout(context).apply { flexDirection = FlexDirection.COLUMN }
        Pair(f, f)
      }
    }
  }

  private fun setProp(id: Long, key: String, value: Any?) {
    val e = entries[id] ?: return
    e.props[key] = value
    val v = e.view
    when (key) {
      "text", "title" -> if (v is TextView) v.text = value?.toString() ?: ""
      "value" -> when (v) {
        is EditText -> if (v.text.toString() != value?.toString()) v.setText(value?.toString() ?: "")
        is SwitchCompat -> v.isChecked = value == true
        is SeekBar -> v.progress = ((value as? Number)?.toInt() ?: 0)
      }
      "placeholder" -> if (v is EditText) v.hint = value?.toString()
      "secure" -> if (v is EditText && value == true) v.inputType = android.text.InputType.TYPE_TEXT_VARIATION_PASSWORD or android.text.InputType.TYPE_CLASS_TEXT
      "min" -> if (v is SeekBar) v.min = ((value as? Number)?.toInt() ?: 0)
      "max" -> if (v is SeekBar) v.max = ((value as? Number)?.toInt() ?: 100)
      "src", "source" -> {} // image loading TBD (URI/asset)
      else -> applyStyle(e, key, value)
    }
  }

  // Layout + visual style. Numbers are dp; flex props go on FlexboxLayout params.
  private fun applyStyle(e: Entry, key: String, value: Any?) {
    val v = e.view
    val num = (value as? Number)?.toDouble()
    when (key) {
      "bg", "backgroundColor" -> parseColor(value)?.let { v.setBackgroundColor(it) }
      "color" -> if (v is TextView) parseColor(value)?.let { v.setTextColor(it) }
      "padding" -> num?.let { val p = dp(it); v.setPadding(p, p, p, p) }
      "paddingH" -> num?.let { val p = dp(it); v.setPadding(p, v.paddingTop, p, v.paddingBottom) }
      "paddingV" -> num?.let { val p = dp(it); v.setPadding(v.paddingLeft, p, v.paddingRight, p) }
      "fontSize" -> if (v is TextView) num?.let { v.setTextSize(TypedValue.COMPLEX_UNIT_SP, it.toFloat()) }
      "fontWeight" -> if (v is TextView && (value == "700" || value == "bold" || value == "600")) v.setTypeface(v.typeface, android.graphics.Typeface.BOLD)
      "radius", "borderRadius" -> {} // needs a GradientDrawable background; TBD
      "direction" -> if (v is FlexboxLayout) v.flexDirection = if (value == "row") FlexDirection.ROW else FlexDirection.COLUMN
      "align" -> if (v is FlexboxLayout) v.alignItems = alignItems(value?.toString())
      "justify" -> if (v is FlexboxLayout) v.justifyContent = justify(value?.toString())
      "flexWrap" -> if (v is FlexboxLayout) v.flexWrap = if (value == "wrap") FlexWrap.WRAP else FlexWrap.NOWRAP
      "width" -> setSize(v, num, null)
      "height" -> setSize(v, null, num)
      // flex/gap/margin applied when the child is attached (needs parent params)
    }
  }

  private fun setSize(v: View, w: Double?, h: Double?) {
    val lp = v.layoutParams ?: ViewGroup.LayoutParams(
      ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT,
    )
    if (w != null) lp.width = dp(w)
    if (h != null) lp.height = dp(h)
    v.layoutParams = lp
  }

  private fun connect(id: Long, event: String, cb: Long) {
    val e = entries[id] ?: return
    e.events[event] = cb
    val v = e.view
    when (event) {
      "press" -> v.setOnClickListener { host.pushEvent(id, "press", null) }
      "changeText" -> if (v is EditText) v.addTextChangedListener(object : TextWatcher {
        override fun afterTextChanged(s: Editable?) { host.pushEvent(id, "changeText", JSONObject.quote(s?.toString() ?: "")) }
        override fun beforeTextChanged(s: CharSequence?, a: Int, b: Int, c: Int) {}
        override fun onTextChanged(s: CharSequence?, a: Int, b: Int, c: Int) {}
      })
      "valueChange" -> when (v) {
        is CompoundButton -> v.setOnCheckedChangeListener { _, checked -> host.pushEvent(id, "valueChange", checked.toString()) }
        is SeekBar -> v.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
          override fun onProgressChanged(sb: SeekBar?, p: Int, fromUser: Boolean) { host.pushEvent(id, "valueChange", p.toString()) }
          override fun onStartTrackingTouch(sb: SeekBar?) {}
          override fun onStopTrackingTouch(sb: SeekBar?) {}
        })
      }
    }
  }

  private fun addChild(parentId: Long, childId: Long, index: Int) {
    val p = entries[parentId] ?: return
    val c = entries[childId] ?: return
    (c.view.parent as? ViewGroup)?.removeView(c.view)
    val childLp = flexParams(c)
    if (index in 0 until p.content.childCount) p.content.addView(c.view, index, childLp)
    else p.content.addView(c.view, childLp)
  }

  // Build FlexboxLayout params for a child from its flex-ish props.
  private fun flexParams(c: Entry): ViewGroup.LayoutParams {
    val base = FlexboxLayout.LayoutParams(
      ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT,
    )
    (c.props["flex"] as? Number)?.let { base.flexGrow = it.toFloat() }
    (c.props["flexGrow"] as? Number)?.let { base.flexGrow = it.toFloat() }
    (c.props["width"] as? Number)?.let { base.width = dp(it.toDouble()) }
    (c.props["height"] as? Number)?.let { base.height = dp(it.toDouble()) }
    (c.props["margin"] as? Number)?.let { val m = dp(it.toDouble()); base.setMargins(m, m, m, m) }
    return base
  }

  private fun removeChild(parentId: Long, childId: Long) {
    val p = entries[parentId] ?: return
    val c = entries[childId] ?: return
    p.content.removeView(c.view)
  }

  private fun setRoot(id: Long) {
    val e = entries[id] ?: return
    root.removeAllViews()
    (e.view.parent as? ViewGroup)?.removeView(e.view)
    root.addView(
      e.view,
      FlexboxLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT),
    )
  }

  private fun parseColor(value: Any?): Int? =
    (value as? String)?.let { try { Color.parseColor(it) } catch (t: Throwable) { null } }

  private fun alignItems(v: String?): Int = when (v) {
    "center" -> AlignItems.CENTER
    "flex-end", "end" -> AlignItems.FLEX_END
    "stretch" -> AlignItems.STRETCH
    else -> AlignItems.FLEX_START
  }

  private fun justify(v: String?): Int = when (v) {
    "center" -> JustifyContent.CENTER
    "flex-end", "end" -> JustifyContent.FLEX_END
    "space-between" -> JustifyContent.SPACE_BETWEEN
    "space-around" -> JustifyContent.SPACE_AROUND
    else -> JustifyContent.FLEX_START
  }
}

File: victor/react-native/modules/elpian-widgets/ios/WidgetController.swift
import UIKit

// Maps the VM's rn.op stream to a real UIView tree — no React. One controller per
// VictorSurfaceView (so widget ids are per-surface, matching Android). Flex
// containers use UIStackView (UIKit's native flex analogue of Android's
// FlexboxLayout); the 15 widget KINDs each map to a native UIView. Events set
// native targets that enqueue back to the VM via the host. iOS twin of
// WidgetController.kt — same op vocabulary, same kind→widget mapping.
final class WidgetController {
  private final class Entry {
    let view: UIView
    let content: UIView   // where children attach (== view unless a scroller)
    let kind: String
    var props: [String: Any] = [:]
    var events: [String: Int] = [:]
    init(view: UIView, content: UIView, kind: String) {
      self.view = view; self.content = content; self.kind = kind
    }
  }

  private let root: UIStackView
  private weak var host: VictorSurfaceView?
  private var entries: [Int: Entry] = [:]
  // UIControl/gesture targets keyed by widget id so `press`/value events carry it.
  private var actionProxies: [Int: WidgetActionProxy] = [:]

  init(root: UIStackView, host: VictorSurfaceView) {
    self.root = root
    self.host = host
  }

  // Apply one op message (see NativeWidgetRenderer.send).
  func apply(_ op: [String: Any]) {
    switch op["t"] as? String {
    case "create": create(int(op["id"]), op["cls"] as? String ?? "", op["k"] as? String ?? "view")
    case "set": setProp(int(op["id"]), op["k"] as? String ?? "", op["v"])
    case "connect": connect(int(op["id"]), op["e"] as? String ?? "", int(op["cb"]))
    case "disconnect": entries[int(op["id"])]?.events.removeValue(forKey: op["e"] as? String ?? "")
    case "add": addChild(int(op["p"]), int(op["c"]), op["i"] as? Int ?? -1)
    case "remove": removeChild(int(op["p"]), int(op["c"]))
    case "clear": clearChildren(int(op["p"]))
    case "free": entries.removeValue(forKey: int(op["id"])); actionProxies.removeValue(forKey: int(op["id"]))
    case "root": setRoot(int(op["id"]))
    case "toast": toast(op["m"] as? String ?? "")
    default: break
    }
  }

  private func create(_ id: Int, _ cls: String, _ kind: String) {
    let (view, content) = build(kind)
    entries[id] = Entry(view: view, content: content, kind: kind)
  }

  private func build(_ kind: String) -> (UIView, UIView) {
    switch kind {
    case "text":
      let l = UILabel(); l.numberOfLines = 0; return (l, l)
    case "input":
      let t = UITextField(); t.borderStyle = .roundedRect; return (t, t)
    case "switch": return single(UISwitch())
    case "slider": return single(UISlider())
    case "button", "victorButton":
      let b = UIButton(type: .system); return (b, b)
    case "activity":
      let a = UIActivityIndicatorView(style: .medium); a.startAnimating(); return (a, a)
    case "image":
      let i = UIImageView(); i.contentMode = .scaleAspectFit; return (i, i)
    case "scene3d":
      return single(UIView()) // Godot view is attached by the Godot module
    case "scroll":
      let sv = UIScrollView()
      let inner = column()
      sv.addSubview(inner)
      inner.translatesAutoresizingMaskIntoConstraints = false
      NSLayoutConstraint.activate([
        inner.leadingAnchor.constraint(equalTo: sv.contentLayoutGuide.leadingAnchor),
        inner.trailingAnchor.constraint(equalTo: sv.contentLayoutGuide.trailingAnchor),
        inner.topAnchor.constraint(equalTo: sv.contentLayoutGuide.topAnchor),
        inner.bottomAnchor.constraint(equalTo: sv.contentLayoutGuide.bottomAnchor),
        inner.widthAnchor.constraint(equalTo: sv.frameLayoutGuide.widthAnchor),
      ])
      return (sv, inner)
    default: // view, list, imageBackground, status, refresh
      let f = column(); return (f, f)
    }
  }

  private func single(_ v: UIView) -> (UIView, UIView) { (v, v) }
  private func column() -> UIStackView {
    let s = UIStackView(); s.axis = .vertical; s.alignment = .fill; s.distribution = .fill
    return s
  }

  private func setProp(_ id: Int, _ key: String, _ value: Any?) {
    guard let e = entries[id] else { return }
    e.props[key] = value
    let v = e.view
    switch key {
    case "text", "title":
      if let l = v as? UILabel { l.text = str(value) }
      else if let b = v as? UIButton { b.setTitle(str(value), for: .normal) }
    case "value":
      if let t = v as? UITextField { if t.text != str(value) { t.text = str(value) } }
      else if let s = v as? UISwitch { s.isOn = (value as? Bool) ?? false }
      else if let s = v as? UISlider { s.value = num(value).map(Float.init) ?? 0 }
    case "placeholder": if let t = v as? UITextField { t.placeholder = str(value) }
    case "secure": if let t = v as? UITextField, (value as? Bool) == true { t.isSecureTextEntry = true }
    case "min": if let s = v as? UISlider, let n = num(value) { s.minimumValue = Float(n) }
    case "max": if let s = v as? UISlider, let n = num(value) { s.maximumValue = Float(n) }
    case "src", "source": break // image loading TBD (URL/asset)
    default: applyStyle(e, key, value)
    }
  }

  // Layout + visual style. Numbers are points; flex props go on the stack view.
  private func applyStyle(_ e: Entry, _ key: String, _ value: Any?) {
    let v = e.view
    switch key {
    case "bg", "backgroundColor": color(value).map { v.backgroundColor = $0 }
    case "color": if let l = v as? UILabel { color(value).map { l.textColor = $0 } }
    case "padding": if let s = v as? UIStackView, let n = num(value) {
      s.isLayoutMarginsRelativeArrangement = true
      s.layoutMargins = UIEdgeInsets(top: CGFloat(n), left: CGFloat(n), bottom: CGFloat(n), right: CGFloat(n))
    }
    case "fontSize": if let l = v as? UILabel, let n = num(value) { l.font = l.font.withSize(CGFloat(n)) }
    case "fontWeight":
      if let l = v as? UILabel, value as? String == "700" || value as? String == "bold" || value as? String == "600" {
        l.font = UIFont.boldSystemFont(ofSize: l.font.pointSize)
      }
    case "radius", "borderRadius": if let n = num(value) { v.layer.cornerRadius = CGFloat(n); v.clipsToBounds = true }
    case "direction": if let s = v as? UIStackView { s.axis = (value as? String == "row") ? .horizontal : .vertical }
    case "align": if let s = v as? UIStackView { s.alignment = alignment(value as? String) }
    case "justify": if let s = v as? UIStackView { s.distribution = distribution(value as? String) }
    case "width": if let n = num(value) { fix(v, .width, CGFloat(n)) }
    case "height": if let n = num(value) { fix(v, .height, CGFloat(n)) }
    default: break
    }
  }

  private func fix(_ v: UIView, _ attr: NSLayoutConstraint.Attribute, _ c: CGFloat) {
    v.translatesAutoresizingMaskIntoConstraints = false
    let anchor = attr == .width ? v.widthAnchor : v.heightAnchor
    anchor.constraint(equalToConstant: c).isActive = true
  }

  private func connect(_ id: Int, _ event: String, _ cb: Int) {
    guard let e = entries[id] else { return }
    e.events[event] = cb
    let proxy = actionProxies[id] ?? WidgetActionProxy(id: id, controller: self)
    actionProxies[id] = proxy
    let v = e.view
    switch event {
    case "press":
      if let b = v as? UIButton {
        b.addTarget(proxy, action: #selector(WidgetActionProxy.onPress), for: .touchUpInside)
      } else {
        let tap = UITapGestureRecognizer(target: proxy, action: #selector(WidgetActionProxy.onPress))
        v.isUserInteractionEnabled = true
        v.addGestureRecognizer(tap)
      }
    case "changeText":
      if let t = v as? UITextField {
        t.addTarget(proxy, action: #selector(WidgetActionProxy.onChangeText(_:)), for: .editingChanged)
      }
    case "valueChange":
      if let c = v as? UIControl {
        c.addTarget(proxy, action: #selector(WidgetActionProxy.onValueChange(_:)), for: .valueChanged)
      }
    default: break
    }
  }

  // Called by the action proxies; forward to the surface's app-scoped event queue.
  func fire(_ id: Int, _ event: String, _ argJson: String?) {
    host?.pushEvent(id, event, argJson)
  }

  private func addChild(_ parentId: Int, _ childId: Int, _ index: Int) {
    guard let p = entries[parentId], let c = entries[childId] else { return }
    c.view.removeFromSuperview()
    if let stack = p.content as? UIStackView {
      if index >= 0 && index < stack.arrangedSubviews.count {
        stack.insertArrangedSubview(c.view, at: index)
      } else {
        stack.addArrangedSubview(c.view)
      }
    } else {
      p.content.addSubview(c.view)
    }
  }

  private func removeChild(_ parentId: Int, _ childId: Int) {
    guard let c = entries[childId] else { return }
    c.view.removeFromSuperview()
  }

  private func clearChildren(_ parentId: Int) {
    guard let p = entries[parentId] else { return }
    if let stack = p.content as? UIStackView {
      stack.arrangedSubviews.forEach { $0.removeFromSuperview() }
    } else {
      p.content.subviews.forEach { $0.removeFromSuperview() }
    }
  }

  private func setRoot(_ id: Int) {
    guard let e = entries[id] else { return }
    root.arrangedSubviews.forEach { $0.removeFromSuperview() }
    e.view.removeFromSuperview()
    root.addArrangedSubview(e.view)
  }

  private func toast(_ message: String) {
    host?.showToast(message)
  }

  // --- coercion helpers -------------------------------------------------------
  private func int(_ v: Any?) -> Int { (v as? Int) ?? Int((v as? Double) ?? 0) }
  private func num(_ v: Any?) -> Double? { (v as? Double) ?? (v as? Int).map(Double.init) }
  private func str(_ v: Any?) -> String {
    if let s = v as? String { return s }
    if let v = v { return "\(v)" }
    return ""
  }
  private func color(_ v: Any?) -> UIColor? {
    guard let hex = v as? String else { return nil }
    return WidgetController.parseHexColor(hex)
  }

  private func alignment(_ v: String?) -> UIStackView.Alignment {
    switch v {
    case "center": return .center
    case "flex-end", "end": return .trailing
    case "stretch": return .fill
    default: return .leading
    }
  }
  private func distribution(_ v: String?) -> UIStackView.Distribution {
    switch v {
    case "space-between": return .equalSpacing
    case "space-around": return .equalCentering
    default: return .fill
    }
  }

  static func parseHexColor(_ hex: String) -> UIColor? {
    var s = hex.trimmingCharacters(in: .whitespaces)
    guard s.hasPrefix("#") else { return nil }
    s.removeFirst()
    if s.count == 3 { s = s.map { "\($0)\($0)" }.joined() }
    guard let v = UInt64(s, radix: 16) else { return nil }
    let hasAlpha = s.count == 8
    let r = CGFloat((v >> (hasAlpha ? 24 : 16)) & 0xff) / 255
    let g = CGFloat((v >> (hasAlpha ? 16 : 8)) & 0xff) / 255
    let b = CGFloat((v >> (hasAlpha ? 8 : 0)) & 0xff) / 255
    let a = hasAlpha ? CGFloat(v & 0xff) / 255 : 1
    return UIColor(red: r, green: g, blue: b, alpha: a)
  }
}

// Bridges UIControl/gesture callbacks (which need an ObjC target) to the
// controller, tagging each with its widget id and the event's arg.
final class WidgetActionProxy: NSObject {
  private let id: Int
  private weak var controller: WidgetController?
  init(id: Int, controller: WidgetController) { self.id = id; self.controller = controller }

  @objc func onPress() { controller?.fire(id, "press", nil) }
  @objc func onChangeText(_ field: UITextField) {
    controller?.fire(id, "changeText", jsonString(field.text ?? ""))
  }
  @objc func onValueChange(_ control: UIControl) {
    if let s = control as? UISwitch { controller?.fire(id, "valueChange", s.isOn ? "true" : "false") }
    else if let s = control as? UISlider { controller?.fire(id, "valueChange", "\(s.value)") }
  }

  private func jsonString(_ s: String) -> String {
    let data = try? JSONSerialization.data(withJSONObject: [s])
    guard let data = data, let arr = String(data: data, encoding: .utf8) else { return "\"\"" }
    // arr is ["..."]; strip the array brackets to get the quoted string.
    return String(arr.dropFirst().dropLast())
  }
}
